File: javascript/src/tuple/Monuple.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Set} from "@joookiwi/type"

import {AbstractTuple}       from "./AbstractTuple"
import {CollectionHolderOf1} from "../CollectionHolderOf1"

export class Monuple<const T = unknown, >
    extends AbstractTuple<T> {

    //#region -------------------- Fields --------------------

    public readonly 0: T
    readonly #value1: T
    #array?: readonly [T,]
    #set?: Set<T>
    #collection?: CollectionHolderOf1<T>
    readonly #hasNull: boolean
    readonly #hasNoNulls: boolean

    //#endregion -------------------- Fields --------------------
    //#region -------------------- Constructor --------------------

    public constructor(value: T,) {
        super()
        this.#hasNoNulls = !(this.#hasNull = (this[0] = this.#value1 = value) == null)
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    /** The value passed through the {@link constructor} held by this instance */
    public get value1(): T { return this.#value1 }


    public override get size(): 1 { return 1 }

    public override get isEmpty(): false { return false }
    public override get isNotEmpty(): true { return true }

    public override get hasExactly1Element(): true { return true }
    public override get hasAtMost1Element(): true { return true }

    public override get hasAtLeast2Elements(): false { return false }
    public override get hasExactly2Elements(): false { return false }
    public override get hasAtMost2Elements(): true { return true }


    public override get hasNull(): boolean { return this.#hasNull }
    public override get hasNoNulls(): boolean { return this.#hasNoNulls }

    public override get hasDuplicate(): false { return false }
    public override get hasNoDuplicates(): true { return true }


    public override toArray(): readonly [T,] { return this.#array ??= Object.freeze([this.value1,],) }

    public override toSet(): Set<T> { return this.#set ??= Object.freeze(new Set([this.value1,],),) }

    public override toCollection(): CollectionHolderOf1<T> { return this.#collection ??= new CollectionHolderOf1(this.value1,) }

    public override toString(): string {
        const value1 = this.value1
        return `Monuple (${value1 == null ? '' : value1})`
    }

    //#endregion -------------------- Methods --------------------

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/tuple/Monuple.java
package joookiwi.collection.java.tuple;

import joookiwi.collection.java.CollectionHolderOf1;
import joookiwi.collection.java.method.ArrayCreator;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.MustBeInvokedByOverriders;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_NEW_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;
import static joookiwi.collection.java.method.AsString.asString;

/// An instance of [Tuple] with only a single value from its `constructor`
///
/// @param <T> The type
@NotNullByDefault
public class Monuple<T extends @Nullable Object>
        extends AbstractTuple<T>
        implements Tuple<T> {

    //#region -------------------- Fields --------------------

    private final T __value1;

    private T @Nullable [] __array;

    //#endregion -------------------- Fields --------------------
    //#region -------------------- Constructor --------------------

    public Monuple(final T value) {
        super();
        __value1 = value;
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    /// The value passed through the `constructor` held by this instance
    public T value1() { return __value1; }

    //#region -------------------- Size methods --------------------

    @Override public @Range(from = 1, to = 1) int size() { return 1; }

    @Contract(ALWAYS_FALSE_0) @Override public boolean isEmpty() { return false; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean isNotEmpty() { return true; }

    @Contract(ALWAYS_TRUE_0) @Override public boolean hasExactly1Element() { return true; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean hasAtMost1Element() { return true; }

    @Contract(ALWAYS_FALSE_0) @Override public boolean hasAtLeast2Elements() { return false; }
    @Contract(ALWAYS_FALSE_0) @Override public boolean hasExactly2Elements() { return false; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean hasAtMost2Elements() { return true; }

    //#endregion -------------------- Size methods --------------------
    //#region -------------------- Has ‥ --------------------

    @Override public boolean hasNull() {
        return value1() == null;
    }

    @Override public boolean hasNoNulls() {
        return value1() != null;
    }


    @Contract(ALWAYS_FALSE_0) @Override public boolean hasDuplicate() {
        return false;
    }

    @Contract(ALWAYS_TRUE_0) @Override public boolean hasNoDuplicates() {
        return true;
    }

    //#endregion -------------------- Has ‥ --------------------
    //#region -------------------- Conversion methods --------------------

    @Override public T[] toArray() {
        final var value = __array;
        if (value != null)
            return value;
        return __array = ArrayCreator.Array(value1());
    }

    @Override public CollectionHolderOf1<T> toCollection() { return new CollectionHolderOf1<>(value1()); }

    @Override public String toString() { return "Monuple (" + asString(value1()) + ')'; }

    //#endregion -------------------- Conversion methods --------------------
    //#region -------------------- Cloneable methods --------------------

    @MustBeInvokedByOverriders
    @Contract(ALWAYS_NEW_0)
    @Override public Monuple<T> clone() { return (Monuple<T>) super.clone(); }

    //#endregion -------------------- Cloneable methods --------------------

    //#endregion -------------------- Methods --------------------

}
